File: pipeline/lib/worker-jobs.ts
/** CPU/memory worker cap shared by Node square shards. */

import { availableParallelism } from 'node:os'
import { readFileSync } from 'node:fs'
import { resolve, dirname } from 'node:path'

const CGROUP_ROOT = '/sys/fs/cgroup'

export function cpuJobs(): number {
  try {
    return availableParallelism()
  } catch {
    return 1
  }
}

export function availableMemoryBytes(): number {
  return cgroupMemoryMax() ?? hostAvailableBytes()
}

export function fitJobs(requested: number, bytesPerWorker: number, memoryBytes = availableMemoryBytes()): number {
  if (!Number.isInteger(requested) || requested < 1) throw new Error('--jobs must be >= 1')
  if (!Number.isInteger(bytesPerWorker) || bytesPerWorker < 1) throw new Error('bytes_per_worker must be >= 1')
  if (!Number.isInteger(memoryBytes) || memoryBytes < 1) throw new Error('memory limit must be >= 1')
  return Math.max(1, Math.min(requested, Math.floor(memoryBytes / bytesPerWorker)))
}

function cgroupMemoryMax(): number | undefined {
  try {
    const text = readFileSync('/proc/self/cgroup', 'utf8')
    const line = text.split('\n').find(row => row.startsWith('0::'))
    if (!line) return undefined
    let path = resolve(CGROUP_ROOT, line.slice(3).replace(/^\//, ''))
    while (true) {
      try {
        const raw = readFileSync(resolve(path, 'memory.max'), 'utf8').trim()
        if (raw !== 'max') {
          const bytes = Number(raw)
          if (Number.isFinite(bytes)) return bytes
        }
      } catch {
        // Walk toward the root until a numeric limit exists.
      }
      if (path === CGROUP_ROOT) return undefined
      const parent = dirname(path)
      if (parent === path) return undefined
      path = parent
    }
  } catch {
    return undefined
  }
}

function hostAvailableBytes(): number {
  try {
    const line = readFileSync('/proc/meminfo', 'utf8').split('\n').find(row => row.startsWith('MemAvailable:'))
    const kilobytes = line?.split(/\s+/)[1]
    if (kilobytes) return Number(kilobytes) * 1024
  } catch {
    // Fall through to os.totalmem-equivalent via /proc.
  }
  const total = readFileSync('/proc/meminfo', 'utf8').split('\n').find(row => row.startsWith('MemTotal:'))
  return Number(total?.split(/\s+/)[1] ?? 0) * 1024
}

File: scripts/lib/worker_jobs.py
"""Process-pool size: all CPUs, or fewer when the cgroup/host cannot hold them."""

from pathlib import Path
import os

_CGROUP_ROOT = Path('/sys/fs/cgroup')


def cpu_jobs():
    try:
        return len(os.sched_getaffinity(0)) or 1
    except (AttributeError, OSError):
        return os.cpu_count() or 1


def available_memory_bytes():
    """Cgroup `memory.max` when finite, otherwise host MemAvailable."""
    cgroup = _cgroup_memory_max()
    if cgroup is not None:
        return cgroup
    return _host_available_bytes()


def fit_jobs(requested, bytes_per_worker, memory_bytes=None):
    """Upper-bound `requested` workers by memory. `--jobs` is the request, not a floor."""
    if requested < 1:
        raise ValueError('--jobs must be >= 1')
    if bytes_per_worker < 1:
        raise ValueError('bytes_per_worker must be >= 1')
    available = available_memory_bytes() if memory_bytes is None else memory_bytes
    if available < 1:
        raise ValueError('memory limit must be >= 1')
    return max(1, min(requested, available // bytes_per_worker))


def _cgroup_memory_max():
    try:
        relative = None
        with open('/proc/self/cgroup', encoding='utf-8') as text:
            for line in text:
                line = line.strip()
                if line.startswith('0::'):
                    relative = line[3:].lstrip('/')
                    break
        if relative is None:
            return None
        path = _CGROUP_ROOT / relative
        while True:
            limit = path / 'memory.max'
            if limit.is_file():
                raw = limit.read_text(encoding='utf-8').strip()
                if raw != 'max':
                    return int(raw)
            if path == _CGROUP_ROOT:
                return None
            path = path.parent
    except (OSError, ValueError):
        return None


def _host_available_bytes():
    try:
        with open('/proc/meminfo', encoding='utf-8') as text:
            for line in text:
                if line.startswith('MemAvailable:'):
                    return int(line.split()[1]) * 1024
    except (OSError, ValueError, IndexError):
        pass
    return os.sysconf('SC_PHYS_PAGES') * os.sysconf('SC_PAGE_SIZE')
